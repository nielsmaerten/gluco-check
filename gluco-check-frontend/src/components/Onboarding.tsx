import React from "react";
import { useTranslation } from "react-i18next";
import {
  Card,
  Container,
  Grid,
  makeStyles,
  Typography,
} from "@material-ui/core";

import assistantIcon from "../images/icon-assistant.svg";

import { CarouselProvider, Dot, Slide, Slider } from "pure-react-carousel";
import "pure-react-carousel/dist/react-carousel.es.css";

import { DEFAULT_ONBOARDING_CAROUSEL_INTERVAL } from "../lib/constants";

const useStyles = makeStyles((theme) => ({
  gridContainer: {
    maxWidth: "500px",
  },
  slideContainer: {
    padding: theme.spacing(2),
    margin: theme.spacing(1),
  },
  callResponseContainer: {
    height: "100%",
    maxHeight: "300px",
    width: "100%",
    "& .assistantBubble img": {
      width: "54px",
      height: "54px",
      display: "block",
      paddingBottom: theme.spacing(2),
      float: "right",
    },
    "& .assistantBubble-question, & .assistantBubble-response": {
      maxWidth: "70%",
    },
    "& .assistantBubble p": {
      clear: "both",
      marginBottom: theme.spacing(2),
      padding: "14px 30px",
      borderRadius: "62px",
      borderWidth: "2px",
      borderStyle: "solid",
    },
    "& .assistantBubble p:last-of-type": {
      marginBottom: "0",
    },
    "& .assistantBubble-question": {
      textAlign: "start",
      alignSelf: "flex-start",
    },
    "& .assistantBubble-question p": {
      backgroundColor: theme.palette.grey[300],
      borderColor: theme.palette.grey[300],
      color: theme.palette.grey[900],
    },
    "& .assistantBubble-response": {
      textAlign: "end",
      alignSelf: "flex-end",
    },
    "& .assistantBubble-response p": {
      borderColor: "#e8eaed",
      textAlign: "inline-end",
    },
    [theme.breakpoints.down("xs")]: {
      "& .assistantBubble p": {
        padding: "7px 15px",
        borderRadius: "20px",
        fontSize: ".9rem",
        lineHeight: "1.1rem",
      },
      "& .assistantBubble img": {
        margin: "0",
        paddingBottom: theme.spacing(1),
      },
    },
  },
  carousel: {
    minWidth: "80vw",
    [theme.breakpoints.down("sm")]: {
      minWidth: "98vw",
    },
    "& .carousel__dot": {
      marginRight: theme.spacing(1),
    },
    "& .carousel__dot:last-of-type": {
      marginRight: "0",
    },
  },
  carouselSliderContainer: {
    width: "100%",
  },
  carouselSlider: {
    minHeight: "300px",
    [theme.breakpoints.down("xs")]: {
      minHeight: "220px",
    },
  },
  carouselDotContainer: {
    paddingBottom: theme.spacing(2),
  },
  carouselDot: {
    backgroundColor: "transparent",
    borderWidth: "0",
    color: theme.palette.text.primary,
    fontFamily: theme.typography.fontFamily,
    "&.carousel__dot--selected": {
      fontWeight: "bold",
    },
  },
}));

export const questionAndAnswer = (
  containerClass: string,
  innerContainerClass: string,
  question: string,
  answer: string,
  index: number
) => {
  return (
    <Card className={innerContainerClass}>
      <Grid
        className={containerClass}
        container
        direction="column"
        key={index}
        wrap="nowrap"
      >
        <Grid item className="assistantBubble assistantBubble-question">
          <Typography variant="body1" component="p">
            {question}
          </Typography>
        </Grid>
        <Grid item className="assistantBubble assistantBubble-response">
          <img src={assistantIcon} alt="" />
          <Typography variant="body1" component="p">
            {answer}
          </Typography>
        </Grid>
      </Grid>
    </Card>
  );
};

interface QuestionsAndAnswer {
  question: string;
  answer: string;
  label: string;
}

function Onboarding() {
  const classes = useStyles();
  const { t } = useTranslation();

  const questionsAndAnswers: QuestionsAndAnswer[] = [
    {
      question: t("onboarding.q1"),
      answer: t("onboarding.a1"),
      label: t("onboarding.l1"),
    },
    {
      question: t("onboarding.q2"),
      answer: t("onboarding.a2"),
      label: t("onboarding.l2"),
    },
    {
      question: t("onboarding.q3"),
      answer: t("onboarding.a3"),
      label: t("onboarding.l3"),
    },
  ];

  return (
    <CarouselProvider
      className={classes.carousel}
      dragEnabled={false}
      infinite={true}
      interval={DEFAULT_ONBOARDING_CAROUSEL_INTERVAL}
      isPlaying={true}
      naturalSlideHeight={300}
      naturalSlideWidth={600}
      totalSlides={questionsAndAnswers.length}
    >
      <Container className={classes.gridContainer} maxWidth={false}>
        <Grid
          alignItems="center"
          container
          direction="column"
          justify="center"
          spacing={2}
        >
          <Grid item className={classes.carouselSliderContainer}>
            <Slider className={classes.carouselSlider}>
              {questionsAndAnswers.map((item, index) => (
                <Slide index={index} key={`slide-${index}`}>
                  {questionAndAnswer(
                    classes.callResponseContainer,
                    classes.slideContainer,
                    item.question,
                    item.answer,
                    index
                  )}
                </Slide>
              ))}
            </Slider>
          </Grid>
          <Grid item className={classes.carouselDotContainer}>
            {questionsAndAnswers.map((item, index) => (
              <Dot
                className={classes.carouselDot}
                disabled={false}
                key={`carouselDot-${index}`}
                slide={index}
              >
                {item.label}
              </Dot>
            ))}
          </Grid>
        </Grid>
      </Container>
    </CarouselProvider>
  );
}

export default Onboarding;
