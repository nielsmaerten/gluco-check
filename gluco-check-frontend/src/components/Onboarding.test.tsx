import React from "react";
import { cleanup, render } from "@testing-library/react";
import { axe } from "jest-axe";
import Onboarding, { questionAndAnswer } from "./Onboarding";

jest.mock("react-i18next", () => ({
  useTranslation: () => {
    return {
      t: jest.fn().mockImplementation((i) => {
        return i;
      }),
    };
  },
}));

jest.mock("pure-react-carousel", () => {
  return {
    __esModule: true,
    CarouselProvider: (props: any) => {
      return <>{props.children}</>;
    },
    Slider: (props: any) => {
      return <>{props.children}</>;
    },
    Slide: (props: any) => {
      return <>{props.children}</>;
    },
    Dot: (props: any) => {
      return <>{props.children}</>;
    },
  };
});

afterEach(cleanup);

describe("Onboarding", () => {
  it("renders the component", () => {
    const { container } = render(<Onboarding />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has no axe violations", async () => {
    const { container } = render(<Onboarding />);
    expect(await axe(container)).toHaveNoViolations();
  });

  describe("questionAndAnswer", () => {
    it("renders the subcomponent", () => {
      expect(
        questionAndAnswer(
          "outerClass",
          "innerClass",
          "What it is",
          "It is good",
          1
        )
      ).toMatchSnapshot();
    });
    it("renders the subcomponent with a long answer string", () => {
      expect(
        questionAndAnswer(
          "outerClass",
          "innerClass",
          "What it is",
          "It is really really really really really good",
          1
        )
      ).toMatchSnapshot();
    });
  });
});
