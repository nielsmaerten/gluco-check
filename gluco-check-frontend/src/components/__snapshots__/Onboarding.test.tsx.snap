// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding questionAndAnswer renders the subcomponent 1`] = `
<WithStyles(ForwardRef(Card))
  className="innerClass"
>
  <WithStyles(ForwardRef(Grid))
    className="outerClass"
    container={true}
    direction="column"
    wrap="nowrap"
  >
    <WithStyles(ForwardRef(Grid))
      className="assistantBubble assistantBubble-question"
      item={true}
    >
      <WithStyles(ForwardRef(Typography))
        component="p"
        variant="body1"
      >
        What it is
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="assistantBubble assistantBubble-response"
      item={true}
    >
      <img
        alt=""
        src="icon-assistant.svg"
      />
      <WithStyles(ForwardRef(Typography))
        component="p"
        variant="body1"
      >
        It is good
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Card))>
`;

exports[`Onboarding questionAndAnswer renders the subcomponent with a long answer string 1`] = `
<WithStyles(ForwardRef(Card))
  className="innerClass"
>
  <WithStyles(ForwardRef(Grid))
    className="outerClass"
    container={true}
    direction="column"
    wrap="nowrap"
  >
    <WithStyles(ForwardRef(Grid))
      className="assistantBubble assistantBubble-question"
      item={true}
    >
      <WithStyles(ForwardRef(Typography))
        component="p"
        variant="body1"
      >
        What it is
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="assistantBubble assistantBubble-response"
      item={true}
    >
      <img
        alt=""
        src="icon-assistant.svg"
      />
      <WithStyles(ForwardRef(Typography))
        component="p"
        variant="body1"
      >
        It is really really really really really good
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Card))>
`;

exports[`Onboarding renders the component 1`] = `
<div
  class="MuiContainer-root makeStyles-gridContainer-1"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root makeStyles-carouselSliderContainer-5 MuiGrid-item"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-slideContainer-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root makeStyles-callResponseContainer-3 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap"
        >
          <div
            class="MuiGrid-root assistantBubble assistantBubble-question MuiGrid-item"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.q1
            </p>
          </div>
          <div
            class="MuiGrid-root assistantBubble assistantBubble-response MuiGrid-item"
          >
            <img
              alt=""
              src="icon-assistant.svg"
            />
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.a1
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-slideContainer-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root makeStyles-callResponseContainer-3 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap"
        >
          <div
            class="MuiGrid-root assistantBubble assistantBubble-question MuiGrid-item"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.q2
            </p>
          </div>
          <div
            class="MuiGrid-root assistantBubble assistantBubble-response MuiGrid-item"
          >
            <img
              alt=""
              src="icon-assistant.svg"
            />
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.a2
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-slideContainer-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root makeStyles-callResponseContainer-3 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap"
        >
          <div
            class="MuiGrid-root assistantBubble assistantBubble-question MuiGrid-item"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.q3
            </p>
          </div>
          <div
            class="MuiGrid-root assistantBubble assistantBubble-response MuiGrid-item"
          >
            <img
              alt=""
              src="icon-assistant.svg"
            />
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              onboarding.a3
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-carouselDotContainer-7 MuiGrid-item"
    >
      onboarding.l1
      onboarding.l2
      onboarding.l3
    </div>
  </div>
</div>
`;
